### ISSUE TITLE
Published transforms are interrupted

### ISSUE THREAD

**Issue by gaschler** [googlecartographer/cartographer_ros#692]
Expected behavior:
Node publishes tf topic every ~10 ms.

Observes behavoir:
Published tf topic is sometimes interrupted for ~100 ms.

How to reproduce:
Execute the following launch file and visualize the timestamps of the generated bag file.
```
<launch>
  <param name="/use_sim_time" value="true" />
  <include file="$(find cartographer_ros)/launch/backpack_2d.launch" />
  <node name="rviz" pkg="rviz" type="rviz"
      args="-d $(find cartographer_ros)/configuration_files/demo_2d.rviz" />
  <node name="playbag" pkg="rosbag" type="play"
      required="true"
      args="--clock $(env HOME)/Downloads/b2-2016-04-27-12-31-41.bag
          --duration=20" />
  <node name="record_tf" pkg="rosbag" type="record"
      required="true"
      args="--output-name=$(env HOME)/Downloads/tf-reproduce.bag
          /tf" />
</launch>
```

```
rqt_bag ~/Downloads/tf-reproduce.bag
```

![screenshot from 2018-01-29 12 56 34](https://user-images.githubusercontent.com/339437/35509396-4b928b18-04f4-11e8-9605-b34a0f93c4bf.png)

**Comment by MichaelGrupp** [comment]
FYI: Could reproduce on my machine.

![screenshot from 2018-01-29 17-02-13](https://user-images.githubusercontent.com/8985495/35520037-2ee603f6-0516-11e8-8abd-f005e767927f.png)

Just a wild guess: maybe a threading / lock issue? Some optimization / other background task?

**Comment by gaschler** [comment]
I did a lot of debugging and my conclusion is this is *intended behavior* of `rosbag play --clock`.
With respect to wall clock time, the largest gap between two `Node::PublishTrajectoryStates` was only 0.020 s, so we publish "odom" in time.

However, the simulated time topic "clock" published by `rosbag play` has exactly these large jumps.
The last message before such jump publishes a simulated timestamp in the future, and then nodes will work in this simulated time, including `rosbag record`.

![screenshot from 2018-01-30 13 37 46](https://user-images.githubusercontent.com/339437/35566820-13ab4be8-05c3-11e8-8297-69af7eac9e8d.png)

Long story short, this does not indicate anything related to real hardware, it only shows we cannot use rosbag (like this) to reproduce timing behavior.

Closing as working as intended.

**Comment by ojura** [comment]
In that case, the offline node is probably affected by this as well. It behaves similarly to rosbag play, and the stamps used there for publishing the clock are serialization timestamps, and _not_ header timestamps (i.e. sensor measurement timestamps):
https://github.com/googlecartographer/cartographer_ros/blob/2538ac3e45ccaf553e956bbb7e745c26008460bf/cartographer_ros/cartographer_ros/offline_node.cc#L279-L280

So, if your bag recording is jumpy (so not the sensor measurements' header stamps, which can be fine, but the serialization times), you get jumpy behaviour when using `sim_time = true`.

**Comment by gaschler** [comment]
The input bag [b2-2016-04-27-12-31-41.bag](https://storage.googleapis.com/cartographer-public-data/bags/backpack_2d/b2-2016-04-27-12-31-41.bag) is not jumpy (neither serialization nor headers).

Thanks for pointing out that offline_node will be similar in this case.
